src/git_credential_msal/main.py

```diff
@@ -147,11 +161,23 @@ def get_msal_cache(name: str) -> SerializableTokenCache:
 def put_msal_cache(name: str, cache: SerializableTokenCache):
     if cache.has_state_changed:
         data = cache.serialize()
-        keyring.set_password("system", name, data)
+        try:
+            keyring.set_password("system", name, data)
+        except keyring.errors.NoKeyringError:
+            os.makedirs(cache_dir, exist_ok=True)
+            msal_cache_name = f"msal_cache_{name}"
+            msal_cache_path = os.path.join(cache_dir, msal_cache_name)
+            with open(msal_cache_path, "w") as f:
+                f.write(data)
```

Comment:
This will automatically write secrets to the disk unencrypted with no user intervention, right? I'm not a huge fan of that idea. Would it make sense to print an error instead and refuse to write the secrets unless users pass an additional flag on the command line, or maybe set an additional git config flag? That could also silence the warning if users are explicitly opting in.

Reply:
I also was not a huge fan of this idea. However, I wanted to come up with a solution that would satisfy the use case of headless machine users. I originally authored an encrypted solution but I wanted to minimize introducing more external dependencies.

https://github.com/Binary-Eater/git-credential-msal/pull/9#issuecomment-2182322510

I like your suggestion. Will go forward with it.

Reply:
I feel like I came up with a better solution security-wise by depending on other credential helpers (especially the git-credential-cache helper). Let me know what you think.

Reply:
No refresh token caching in this model, but I think that's fair in an ecosystem where the user chooses not the use a keyring provider. Using the cache helper makes it so such a use case only involves one auth dance in a partial clone operation in theory.

Reply:
Right, but isn't any later use going to require another auth dance? That seems pretty user unfriendly.

Given that git-credential-store doesn't seem well suited to this, it does seem worth having git-credential-msal be able to store the refresh token, with an opt-in if it's insecure.

Reply:
https://github.com/Binary-Eater/git-credential-msal/pull/9/commits/1046d3d623c09587afb894fd4c0e0614672c7623

Reply:
I had some iterations from testing :joy:, https://github.com/Binary-Eater/git-credential-msal/commit/e74b662c84e516a8f5159537428299720fd23038